File: mobile/packages/footprint-react-native/src/utils/create-url.ts
import type { FootprintVerifyProps } from '../footprint.types';

const encode = (obj?: Record<string, unknown>): string => {
  return obj && Object.keys(obj).length
    ? encodeURIComponent(JSON.stringify(obj))
    : '';
};

const createUrl = ({
  appearance,
  l10n,
  redirectUrl,
  token,
}: {
  appearance?: FootprintVerifyProps['appearance'];
  l10n?: FootprintVerifyProps['l10n'];
  redirectUrl?: string;
  token: string;
}) => {
  const searchParams = new URLSearchParams();

  if (redirectUrl) {
    searchParams.append('redirect_url', redirectUrl);
  }

  if (appearance) {
    const variables = encode(appearance.variables);
    const rules = encode(appearance.rules);

    if (variables) {
      searchParams.append('variables', variables);
    }
    if (rules) {
      searchParams.append('rules', rules);
    }
    if (appearance.fontSrc) {
      searchParams.append('font_src', appearance.fontSrc);
    }
    if (l10n?.language) {
      searchParams.append('lng', l10n.language);
    }
  }

  return `https://id.onefootprint.com?${searchParams.toString()}#${token}`;
};

export default createUrl;

File: mobile/packages/footprint_flutter/lib/utils/create-url.dart
part of "../footprint_flutter.dart";

String _createUrl({
  FootprintAppearance? appearance,
  FootprintL10n? l10n,
  String? redirectUrl,
  required String token,
}) {
  var searchParams = StringBuffer();

  void addParam(String key, String value) {
    if (searchParams.isNotEmpty) {
      searchParams.write('&');
    }
    searchParams.write('$key=$value');
  }

  if (redirectUrl != null) {
    addParam('redirect_url', redirectUrl);
  }

  if (appearance != null) {
    var variables = _encode(appearance.variables?._toJson());
    var rules = _encode(appearance.rules?._toJson());

    if (variables.isNotEmpty) {
      addParam('variables', variables);
    }
    if (rules.isNotEmpty) {
      addParam('rules', rules);
    }
    if (appearance.fontSrc != null) {
      addParam('font_src', appearance.fontSrc!);
    }
  }

  return 'https://id.onefootprint.com?$searchParams#$token';
}

String _encode(Map<String, dynamic>? obj) {
  return obj != null && obj.isNotEmpty
      ? Uri.encodeComponent(json.encode(obj))
      : '';
}
